File: tests/regression/go_project/service.go
package main

import "fmt"

// Repository is a generic data access interface.
type Repository interface {
	FindByID(id int64) (interface{}, error)
	Save(entity interface{}) (int64, error)
	Delete(id int64) error
	ListAll() []interface{}
}

// UserService manages users.
type UserService struct {
	users  map[int64]*User
	nextID int64
}

// NewUserService creates a new service.
func NewUserService() *UserService {
	return &UserService{users: make(map[int64]*User), nextID: 1}
}

// CreateUser adds a new user.
func (s *UserService) CreateUser(name, email string) (int64, error) {
	user := NewUser(s.nextID, name, email)
	if err := user.Validate(); err != nil {
		return 0, fmt.Errorf("validation failed: %w", err)
	}
	id := user.GetID()
	s.users[id] = user
	s.nextID++
	return id, nil
}

// DeactivateUser marks a user as inactive.
func (s *UserService) DeactivateUser(id int64) error {
	user, ok := s.users[id]
	if !ok {
		return fmt.Errorf("user %d not found", id)
	}
	user.Deactivate()
	return nil
}

// FindByID returns a user by ID.
func (s *UserService) FindByID(id int64) (*User, error) {
	user, ok := s.users[id]
	if !ok {
		return nil, fmt.Errorf("user %d not found", id)
	}
	return user, nil
}

func (s *UserService) generateID() int64 {
	id := s.nextID
	s.nextID++
	return id
}

File: tests/regression/rust_project/src/service.rs
use std::collections::HashMap;

use crate::errors::AppError;
use crate::models::{User, Validatable, Identifiable};

pub struct UserService {
    users: HashMap<u64, User>,
    next_id: u64,
}

pub trait Repository<T> {
    fn find_by_id(&self, id: u64) -> Option<&T>;
    fn save(&mut self, entity: T) -> Result<u64, AppError>;
    fn delete(&mut self, id: u64) -> Result<(), AppError>;
    fn list_all(&self) -> Vec<&T>;
}

impl UserService {
    pub fn new() -> Self {
        UserService { users: HashMap::new(), next_id: 1 }
    }

    pub fn create_user(&mut self, name: String, email: String) -> Result<u64, AppError> {
        let user = User::new(self.next_id, name, email);
        user.validate()?;
        let id = user.id();
        self.users.insert(id, user);
        self.next_id += 1;
        Ok(id)
    }

    pub fn deactivate_user(&mut self, id: u64) -> Result<(), AppError> {
        match self.users.get_mut(&id) {
            Some(user) => {
                user.deactivate();
                Ok(())
            }
            None => Err(AppError::NotFound(format!("user {}", id))),
        }
    }

    fn generate_id(&mut self) -> u64 {
        let id = self.next_id;
        self.next_id += 1;
        id
    }
}

impl Repository<User> for UserService {
    fn find_by_id(&self, id: u64) -> Option<&User> {
        self.users.get(&id)
    }

    fn save(&mut self, entity: User) -> Result<u64, AppError> {
        entity.validate()?;
        let id = entity.id();
        self.users.insert(id, entity);
        Ok(id)
    }

    fn delete(&mut self, id: u64) -> Result<(), AppError> {
        self.users.remove(&id)
            .map(|_| ())
            .ok_or_else(|| AppError::NotFound(format!("user {}", id)))
    }

    fn list_all(&self) -> Vec<&User> {
        self.users.values().collect()
    }
}
